mm2src/coins/tendermint/tendermint_tx_history_v2.rs

```diff
@@ -469,14 +506,21 @@ where
                             transfer_event_type: TransferEventType::default(),
                         };
 
-                        if index != 0 {
-                            // If previous message is htlc related, that means current transfer
-                            // addresses will be wrong.
-                            if let Some(prev_event) = tx_events.get(index - 1) {
-                                if [CREATE_HTLC_EVENT, CLAIM_HTLC_EVENT].contains(&prev_event.kind.as_str()) {
-                                    read_real_htlc_addresses(&mut tx_details, prev_event);
-                                }
-                            };
+                        // For HTLC transactions, the sender and receiver addresses in the "transfer" event will be incorrect.
+                        // Use `read_real_htlc_addresses` to handle them properly.
+                        if let Some(htlc_event) = tx_events
+                            .iter()
+                            .find(|e| [CREATE_HTLC_EVENT, CLAIM_HTLC_EVENT].contains(&e.kind.as_str()))
+                        {
+                            read_real_htlc_addresses(&mut tx_details, htlc_event);
+                        }
+
+                        // For IBC transactions, the sender and receiver addresses in the "transfer" event will be incorrect.
+                        // Use `read_real_ibc_addresses` to handle them properly.
+                        if let Some(ibc_event) = tx_events.iter().find(|e| {
+                            [IBC_SEND_EVENT, IBC_RECEIVE_EVENT, IBC_NFT_RECEIVE_EVENT].contains(&e.kind.as_str())
+                        }) {
+                            read_real_ibc_addresses(&mut tx_details, ibc_event);
```

Comment:
Q: so this is an either-or relationship? not that a tx can contain IBC send/recv + HTLC events at the same time?

also, is having the ibc or htlc event *anywhere* in the events list a valid check to change the from/to addresses for all tx_details? we seemed to fine-grain this in the past by checking only if the previous event was an HTLC event but we no longer check it this way.

Reply:
> Q: so this is an either-or relationship? not that a tx can contain IBC send/recv + HTLC events at the same time?

It can't be.

> also, is having the ibc or htlc event anywhere in the events list a valid check to change the from/to addresses for all tx_details? we seemed to fine-grain this in the past by checking only if the previous event was an HTLC event but we no longer check it this way.

It's fine. Previous check was also correct, I just improved it.